= #17342 Make 2.4 binary compatible with 2.3

patriknw opened the pull request:
Not ready (will fail), but a good starting pointing for splitting up the remaining work.

patriknw committed:
=act #17239 Add akka.japi.function with variance

Needed by Akka Streams. Old functions placed in akka.japi will be deprecated
in 2.4

(cherry picked from commit 99628f408295070848af6c23b1d722057069e660)

+act #17392 Include generated japi.function from akka-stream

* add boilerplate plugin
* make them Serializable to be able to grab line number for Java 8 lambdas

(cherry picked from commit d5950a13d2f123d2101d56f0a8a86a2097dda8e1)

patriknw committed:
= #17342 Make 2.4 binary compatible with 2.3

(cherry picked from commit 89af8bdb90c98759a3459c1468fa1a861e0f81da)

* remove final identifier in serializers

i* revert/deprecate ProtobufSerializer.ARRAY_OF_BYTE_ARRAY

* adding back compatible empty constructor in serializers

* make FSM.State compatible

* add back ActorPath.ElementRegex

* revert SocketOption changes and add SocketOptionV2
  see a6d3704ef64142c810ca881314261a7ac63caa26

* problem filter for ActorSystem and ActorPath

* problem filter for ByteString

* problem filter for deprecated Timeout methods

* BalancingPool companion

* ask

* problem filter for ActorDSL

* event bus

* exclude hasSubscriptions

* exclude some problems in testkit

* boundAddress and addressFromSocketAddress

* Pool nrOfInstances

* PromiseActorRef

* check with 2.3.9

* migration guide note

* explicit exclude of final class problems

patriknw commented:
The commits marked with ! are:

    69387bb 2014-10-28 | !act #16125 Timeout Calls through CircuitBreaker. (origin/pr/16161) [Martin Eigenbrodt]
    3f12ef2 2014-09-05 | !clu #15042 useRole restriction on local node is now respected This is an API breaking change if someone implemented their own Routers. The change is required because the router must know if the local routees should be started or not so it has to check the roles of the cluster member (the local one). We could delay this decision of starting local routees, but that would allow messages to be dead-letter-ed (bad). (origin/pr/15822) [Konrad 'ktoso' Malawski]
    325e05e 2014-08-19 | !act #15626 expose DatagramChannel creation in DatagramChannelCreator (origin/pr/15627) [Martynas Mickevicius]
    a6d3704 2013-09-08 | !act #13490 Changed the callback to SocketOption to accept a channel instead of a Socket, this allows for using the nio features. (origin/pr/15568) [Michael R. Maletich]
    c046cdf 2014-02-22 | !act,doc #3893 Removed isTerminated checks from ActorClassification (origin/pr/15024) [Konrad Malawski]
    826cc74 2014-03-11 | !tes #2971 Make TestKit.remaining throw AssertionError outside of within (origin/pr/2005) [dario.rexin]

patriknw commented:
remaining mimaReportBinaryIssues:

    [info] akka-actor: found 47 potential binary incompatibilities (filtered 240)
    [error]  * the type hierarchy of class akka.actor.ChildActorPath has changed in new version. Missing types {akka.actor.ActorPath}
    [error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("akka.actor.ChildActorPath")
    [error]  * method ElementRegex()scala.util.matching.Regex in object akka.actor.ActorPath does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.actor.ActorPath.ElementRegex")
    [error]  * method this(java.lang.Object,java.lang.Object,scala.Option,scala.Option,scala.collection.immutable.List)Unit in class akka.actor.FSM#State does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.actor.FSM#State.this")
    [error]  * abstract method logFilter()akka.event.LoggingFilter in class akka.actor.ExtendedActorSystem does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.actor.ExtendedActorSystem.logFilter")
    [error]  * abstract method terminate()scala.concurrent.Future in class akka.actor.ActorSystem does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.actor.ActorSystem.terminate")
    [error]  * abstract method whenTerminated()scala.concurrent.Future in class akka.actor.ActorSystem does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.actor.ActorSystem.whenTerminated")
    [error]  * the type hierarchy of class akka.actor.RootActorPath has changed in new version. Missing types {akka.actor.ActorPath}
    [error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("akka.actor.RootActorPath")
    [error]  * method apply(java.lang.Object,java.lang.Object,scala.Option,scala.Option,scala.collection.immutable.List)akka.actor.FSM#State in object akka.actor.FSM#State does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.actor.FSM#State.apply")
    [error]  * declaration of trait akka.actor.ActorPath has changed to class akka.actor.ActorPath in new version; changing trait to class breaks client code
    [error]    filter with: ProblemFilters.exclude[IncompatibleTemplateDefProblem]("akka.actor.ActorPath")
    [error]  * method boss()akka.actor.RepointableActorRef in class akka.actor.ActorDSL#Extension does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.actor.ActorDSL#Extension.boss")
    [error]  * method afterConnect(java.nio.channels.SocketChannel)Unit in trait akka.io.Inet#SocketOption does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.io.Inet#SocketOption.afterConnect")
    [error]  * method afterConnect(java.nio.channels.ServerSocketChannel)Unit in trait akka.io.Inet#SocketOption does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.io.Inet#SocketOption.afterConnect")
    [error]  * method afterConnect(java.nio.channels.DatagramChannel)Unit in trait akka.io.Inet#SocketOption does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.io.Inet#SocketOption.afterConnect")
    [error]  * method beforeBind(java.nio.channels.SocketChannel)Unit in trait akka.io.Inet#SocketOption does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.io.Inet#SocketOption.beforeBind")
    [error]  * method beforeBind(java.nio.channels.ServerSocketChannel)Unit in trait akka.io.Inet#SocketOption does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.io.Inet#SocketOption.beforeBind")
    [error]  * method beforeBind(java.nio.channels.DatagramChannel)Unit in trait akka.io.Inet#SocketOption does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.io.Inet#SocketOption.beforeBind")
    [error]  * method fromActorSystem(akka.actor.ActorSystem)akka.event.EventStream in object akka.event.EventStream does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.EventStream.fromActorSystem")
    [error]  * synthetic method <init>$default$1()Boolean in object akka.event.EventStream does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.EventStream.<init>$default$1")
    [error]  * method this(Boolean)Unit in class akka.event.EventStream does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.EventStream.this")
    [error]  * method hasSubscriptions(java.lang.Object)Boolean in trait akka.event.SubchannelClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.SubchannelClassification.hasSubscriptions")
    [error]  * method dissociate(akka.actor.ActorRef)scala.collection.immutable.Iterable in trait akka.event.ActorClassification does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.dissociate")
    [error]  * synthetic method akka$event$ActorClassification$$mappings()java.util.concurrent.ConcurrentHashMap in trait akka.event.ActorClassification has now a different result type; was: java.util.concurrent.ConcurrentHashMap, is now: java.util.concurrent.atomic.AtomicReference
    [error]    filter with: ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.event.ActorClassification.akka$event$ActorClassification$$mappings")
    [error]  * synthetic method akka$event$ActorClassification$_setter_$akka$event$ActorClassification$$mappings_=(java.util.concurrent.ConcurrentHashMap)Unit in trait akka.event.ActorClassification's type has changed; was (java.util.concurrent.ConcurrentHashMap)Unit, is now: (java.util.concurrent.atomic.AtomicReference)Unit
    [error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.event.ActorClassification.akka$event$ActorClassification$_setter_$akka$event$ActorClassification$$mappings_=")
    [error]  * method unsubscriber()akka.actor.ActorRef in trait akka.event.ActorClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.unsubscriber")
    [error]  * method dissociate(akka.actor.ActorRef)Unit in trait akka.event.ActorClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.dissociate")
    [error]  * method unregisterFromUnsubscriber(akka.actor.ActorRef,Int)Boolean in trait akka.event.ActorClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.unregisterFromUnsubscriber")
    [error]  * method registerWithUnsubscriber(akka.actor.ActorRef,Int)Boolean in trait akka.event.ActorClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.registerWithUnsubscriber")
    [error]  * synthetic method akka$event$ActorClassification$$mappings()java.util.concurrent.atomic.AtomicReference in trait akka.event.ActorClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.akka$event$ActorClassification$$mappings")
    [error]  * synthetic method akka$event$ActorClassification$_setter_$akka$event$ActorClassification$$mappings_=(java.util.concurrent.atomic.AtomicReference)Unit in trait akka.event.ActorClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.akka$event$ActorClassification$_setter_$akka$event$ActorClassification$$mappings_=")
    [error]  * method system()akka.actor.ActorSystem in trait akka.event.ActorClassification does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.ActorClassification.system")
    [error]  * method this()Unit in class akka.event.japi.ActorEventBus does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.event.japi.ActorEventBus.this")
    [error]  * method ask(java.lang.Object,akka.util.Timeout)scala.concurrent.Future in class akka.pattern.AskableActorRef does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorRef.ask")
    [error]  * method ?(java.lang.Object,akka.util.Timeout)scala.concurrent.Future in class akka.pattern.AskableActorRef does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorRef.?")
    [error]  * synthetic method ask$extension(akka.actor.ActorRef,java.lang.Object,akka.util.Timeout)scala.concurrent.Future in object akka.pattern.AskableActorRef does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorRef.ask$extension")
    [error]  * synthetic method ?$extension(akka.actor.ActorRef,java.lang.Object,akka.util.Timeout)scala.concurrent.Future in object akka.pattern.AskableActorRef does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorRef.?$extension")
    [error]  * synthetic method ask$extension(akka.actor.ActorSelection,java.lang.Object,akka.util.Timeout)scala.concurrent.Future in object akka.pattern.AskableActorSelection does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorSelection.ask$extension")
    [error]  * synthetic method ?$extension(akka.actor.ActorSelection,java.lang.Object,akka.util.Timeout)scala.concurrent.Future in object akka.pattern.AskableActorSelection does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorSelection.?$extension")
    [error]  * method ask(java.lang.Object,akka.util.Timeout)scala.concurrent.Future in class akka.pattern.AskableActorSelection does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorSelection.ask")
    [error]  * method ?(java.lang.Object,akka.util.Timeout)scala.concurrent.Future in class akka.pattern.AskableActorSelection does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskableActorSelection.?")
    [error]  * method ask(akka.actor.ActorSelection,java.lang.Object,akka.actor.ActorRef,akka.util.Timeout)scala.concurrent.Future in trait akka.pattern.AskSupport does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskSupport.ask")
    [error]  * method ask(akka.actor.ActorRef,java.lang.Object,akka.actor.ActorRef,akka.util.Timeout)scala.concurrent.Future in trait akka.pattern.AskSupport does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskSupport.ask")
    [error]  * deprecated method apply(Long)akka.util.Timeout in object akka.util.Timeout's type has changed; was (Long)akka.util.Timeout, is now: (scala.concurrent.duration.FiniteDuration)akka.util.Timeout
    [error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.util.Timeout.apply")
    [error]  * abstract method byteStringCompanion()akka.util.ByteString#Companion in class akka.util.ByteString does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.util.ByteString.byteStringCompanion")
    [error]  * abstract method writeToOutputStream(java.io.ObjectOutputStream)Unit in class akka.util.ByteString does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.util.ByteString.writeToOutputStream")
    [error]  * deprecated method this(Long)Unit in class akka.util.Timeout's type has changed; was (Long)Unit, is now: (scala.concurrent.duration.FiniteDuration)Unit
    [error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.util.Timeout.this")
    [error]  * the type hierarchy of object akka.routing.BalancingPool has changed in new version. Missing types {scala.runtime.AbstractFunction3}
    [error]    filter with: ProblemFilters.exclude[MissingTypesProblem]("akka.routing.BalancingPool$")
    [error]  * method apply(java.lang.Object,java.lang.Object,java.lang.Object)java.lang.Object in object akka.routing.BalancingPool's type has changed; was (java.lang.Object,java.lang.Object,java.lang.Object)java.lang.Object, is now: (Int,akka.actor.SupervisorStrategy,java.lang.String)akka.routing.BalancingPool
    [error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.routing.BalancingPool.apply")


    [info] akka-testkit: found 5 potential binary incompatibilities (filtered 14)
    [error]  * method testActorName()java.lang.String in trait akka.testkit.TestKitBase does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.testkit.TestKitBase.testActorName")
    [error]  * method remainingOrDefault()scala.concurrent.duration.FiniteDuration in trait akka.testkit.TestKitBase does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.testkit.TestKitBase.remainingOrDefault")
    [error]  * method shutdownActorSystem(akka.actor.ActorSystem,java.lang.Boolean)Unit in class akka.testkit.JavaTestKit does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.testkit.JavaTestKit.shutdownActorSystem")
    [error]  * method shutdownActorSystem(akka.actor.ActorSystem,scala.concurrent.duration.Duration)Unit in class akka.testkit.JavaTestKit does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.testkit.JavaTestKit.shutdownActorSystem")
    [error]  * deprecated method dilated(scala.concurrent.duration.Duration,akka.actor.ActorSystem)scala.concurrent.duration.Duration in object akka.testkit.TestKit does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.testkit.TestKit.dilated")


    [info] akka-remote: found 3 potential binary incompatibilities (filtered 214)
    [error]  * method boundAddress()akka.actor.Address in trait akka.remote.transport.Transport does not have a correspondent in old version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.remote.transport.Transport.boundAddress")
    [error]  * method addressFromSocketAddress(java.net.SocketAddress,java.lang.String,java.lang.String,scala.Option)scala.Option in object akka.remote.transport.netty.NettyTransport does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.remote.transport.netty.NettyTransport.addressFromSocketAddress")
    [error]  * method nrOfInstances()Int in class akka.remote.routing.RemoteRouterConfig does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.remote.routing.RemoteRouterConfig.nrOfInstances")


    [info] akka-multi-node-testkit: found 1 potential binary incompatibilities (filtered 31)
    [error]  * synthetic method akka$remote$testkit$MultiNodeSpec$Replacement$$$outer()akka.remote.testkit.MultiNodeSpec in class akka.remote.testkit.MultiNodeSpec#Replacement does not have a correspondent in new version
    [error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("akka.remote.testkit.MultiNodeSpec#Replacement.akka$remote$testkit$MultiNodeSpec$Replacement$$$outer")

patriknw commented on akka-actor/src/main/scala/akka/event/EventBus.scala:
```diff
@@ -180,6 +180,8 @@ trait SubchannelClassification { this: EventBus ⇒
    */
   private[akka] def hasSubscriptions(subscriber: Subscriber): Boolean =
+    // FIXME binary incompatible, but I think it is safe to filter out this problem,
+    //       since it is only called from new functionality in EventStreamUnsubscriber
```
@rkuhn @ktoso please verify my thinking here

ktoso replied:
Yes I remember this method - it's only used internally in this one spot (by the Unsubscriber)

patriknw replied:
the problem is that the method will not exist in user's classes, since this is a trait, until they compile, but I guess that is ok since they will not be using the new EventStreamUnsubscriber in that case anyway

ktoso replied:
Correct, the unsubscriber has to be started somewhere (we do in the ActorSystem for the EventStream), if the user has his own bus they didn't start an unsubscriber - so no-one will call this method

patriknw replied:
thanks for confirming

ktoso commented:
Wohoo!

patriknw commented:
yay!

rkuhn commented:
wow, epic! (now I only need to review this :-) )

rkuhn commented on akka-actor/src/main/scala/akka/actor/FSM.scala:
```diff
@@ -123,17 +123,42 @@ object FSM {
+     * INTERNAL API
+     */
+    private[akka] def copy2(stateName: S = stateName, stateData: D = stateData, timeout: Option[FiniteDuration] = timeout, stopReason: Option[Reason] = stopReason, replies: List[Any] = replies): State[S, D] = {
+      new State(stateName, stateData, timeout, stopReason, replies)
```
Why name this `copy2`? This is a little bit of a trap now, since using `copy` destroys information. Was it impossible to keep the compiler from emitting the `copy` method?

patriknw replied:
yes, I could not make it work with the overridden copy in `SilentState`.
I think I tried to define an exact same signature of `copy` here as the generated one, but that did not work out. I can try again, don't remember exactly the problem.

patriknw replied:
I must have done something stupid before. Works fine now, with defining `copy` instead of `copy2`.

rkuhn commented on akka-actor/src/main/scala/akka/io/Inet.scala:
```diff
@@ -16,38 +19,51 @@ object Inet {
      * the slave socket for servers).
      */
-    def afterConnect(c: ServerSocketChannel): Unit = ()
+    def afterConnect(c: DatagramChannel): Unit = ()
```
This is confusing (well, it was already before your change): this applies only “after bind” but is named “after connect”, while the differentiator in the method signature is the socket/channel dichotomy. Since this did not exist in 2.3 we should rename this to `afterBind` instead.

patriknw replied:
ok, `afterBind` in trait `AfterChannelBind`

rkuhn commented on akka-actor/src/main/scala/akka/io/UdpConnection.scala:
```diff
@@ -67,7 +67,6 @@ private[io] class UdpConnection(udpConn: UdpConnectedExt,
 
   def receive = {
     case registration: ChannelRegistration ⇒
-      options.foreach(_.afterConnect(channel))
```
we may want to add an `afterConnect(DatagramSocket)`, WDYT?

patriknw replied:
I don't know. Adding another trait?
I think this API is a mess, to be honest.

rkuhn replied:
Or collect all new things on `SocketOptionV2`.

rkuhn commented on project/MiMa.scala:
```diff
@@ -17,10 +17,230 @@ object MiMa extends AutoPlugin {
+      // copied everything above from release-2.3 branch
+      
+      // final case classes
+      IgnoreFinalClassProblem, 
```
Hmm, do we really want to suppress all errors related to final classes? This seems risky.

patriknw replied:
it's a lot of case classes, I'll see if I can refine the filter to find out that it is a case class

patriknw replied:
alright, I changed this to explicitly exclude all reported problems. That should prevent us from introducing a new problem later.

rkuhn commented on project/MiMa.scala:
```diff
@@ -17,10 +17,230 @@ object MiMa extends AutoPlugin {
+      ProblemFilters.exclude[MissingMethodProblem]("akka.actor.IndirectActorProducer.UntypedActorFactoryConsumerClass"),
+      FilterAnyProblem("akka.actor.ActorSystemImpl"),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.pattern.AskSupport.ask"),
+      FilterAnyProblem("akka.pattern.PromiseActorRef"),
```
This should probably be refined or go away.

patriknw replied:
strange, I did fix PromiseActorRef, I'll investigate why that is not included here

patriknw replied:
ah, I had not pushed that commit

rkuhn commented on project/MiMa.scala:
```diff
@@ -17,10 +17,230 @@ object MiMa extends AutoPlugin {
+      ProblemFilters.exclude[IncompatibleTemplateDefProblem]("akka.actor.ActorPath"),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.actor.LocalActorRefProvider.terminationPromise"),
+      ProblemFilters.exclude[MissingClassProblem]("akka.actor.UntypedActorFactoryConsumer"),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.actor.FSM#State.copy"),
```
This should also go, the `copy` method may well be used externally.

patriknw replied:
ok, I'll look into that

rkuhn commented on project/MiMa.scala:
```diff
@@ -17,10 +17,230 @@ object MiMa extends AutoPlugin {
+      // class akka.cluster.routing.AdaptiveLoadBalancingRouter does not have a correspondent in new version
+      ProblemFilters.exclude[MissingClassProblem]("akka.cluster.routing.AdaptiveLoadBalancingRouter"),
+      // class akka.cluster.routing.ClusterRouterConfig does not have a correspondent in new version
+      ProblemFilters.exclude[MissingClassProblem]("akka.cluster.routing.ClusterRouterConfig"),
```
Hmm, I guess we cannot automatically verify that the moved versions are identical, but we should keep this in mind.

rkuhn replied:
ah, forget it, these are the deprecation removals.

rkuhn replied:
I’m reading bottom to top here :-)

patriknw replied:
yes, deprecated section

rkuhn commented on project/MiMa.scala:
```diff
@@ -17,10 +17,230 @@ object MiMa extends AutoPlugin {
+      // method remainingOrDefault()scala.concurrent.duration.FiniteDuration in trait akka.testkit.TestKitBase does not have a correspondent in old version
+      ProblemFilters.exclude[MissingMethodProblem]("akka.testkit.TestKitBase.remainingOrDefault"),
+      // synthetic method akka$remote$testkit$MultiNodeSpec$Replacement$$$outer()akka.remote.testkit.MultiNodeSpec in class akka.remote.testkit.MultiNodeSpec#Replacement does not have a correspondent in new version
+      ProblemFilters.exclude[MissingMethodProblem]("akka.remote.testkit.MultiNodeSpec#Replacement.akka$remote$testkit$MultiNodeSpec$Replacement$$$outer"),
```
we need to make it clear that BC does not apply to testkits (and ask about feedback for that)

patriknw replied:
yes, in release notes of M1
I can add something in the migration guide also.

rkuhn commented on project/MiMa.scala:
```diff
@@ -17,10 +17,230 @@ object MiMa extends AutoPlugin {
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor#Batch.blockOn"),
+      ProblemFilters.exclude[FinalMethodProblem]("akka.dispatch.BatchingExecutor#Batch.run"),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor#Batch.akka$dispatch$BatchingExecutor$Batch$$parentBlockContext_="),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor#Batch.this"),
```
It would be great to have a minimum applicable version on these filters, indicating when the change was made. I mean that we expect 2.3.11 to be BC with 2.3.10 in these regards, but now we don’t verify that anymore. @2m, do you think this is doable?

2m replied:
Yes. I think it is doable. https://github.com/typesafehub/migration-manager/issues/65

patriknw commented:
@rkuhn I have incorporated your feedback, you can review the latest commits, starting from PromiseActorRef

patriknw commented on akka-actor/src/main/scala/akka/io/Inet.scala:
```diff
@@ -16,38 +19,65 @@ object Inet {
+     * the slave socket for servers).
+     */
+    // FIXME where to call this one? Is it needed?
+    // def afterConnect(s: ServerSocket): Unit = ()
```
@rkuhn I'm lost here. I renamed afterConnect to afterBind, and added new afterConnect, which is invoked in the UDP case, but I cannot find where to invoke it for TCP. Note that the old SocketOption also has `def afterConnect(s: Socket)`.

rkuhn replied:
AFAIU we already had it in the TCP case, which is why that is on SocketOption, and then we added UDP support later, which is now on SocketOptionV2. So I believe that everything should be okay now, what do you feel is missing?

patriknw replied:
That makes sense. I was probably thinking that there should be one with `ServerSocket` param.

viktorklang committed:
=act - Reinstates the Mailbox optimization reusing the FJT instance but doesn't use FJT.fork() due to unfairness

rkuhn commented:
So, if the UDP afterConnect is solved, we could merge this _major leap_ :-)

patriknw commented:
yes, I'll remove the FIXME regarding the `afterConnect` and squash this

patriknw commented:
@jsureth I have a question for you. Do you think it is alright to change a `sealed trait` to `sealed abstract class`. We have changed [ActorPath](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/ActorPath.scala#L108).

Mima reports:
* the type hierarchy of class akka.actor.ChildActorPath has changed in new version. Missing types {akka.actor.ActorPath}
* the type hierarchy of class akka.actor.RootActorPath has changed in new version. Missing types {akka.actor.ActorPath}
* declaration of trait akka.actor.ActorPath has changed to class akka.actor.ActorPath in new version; changing trait to class breaks client code

The reason we changed this was to get the static forwarders in the companion object.

patriknw commented:
I have rebased and squashed this. Found that I need Viktor's commit for the fork join task, so that PR should be merged first: https://github.com/akka/akka/pull/17444

I have also included the japi functions, cherry-picked as is from release-2.3. No deprecations.

patriknw commented:
merging, if we find anything that is wrong when doing real world testing we will fix that in separate PR

patriknw commented:
Refs #17342

patriknw commented:
ah, forgot my own note about the FJ task, no harm I think that PR was ready to be merged anyway
